mirage/tar_mirage.ml

```diff
@@ -427,9 +455,119 @@ module Make_KV_RW (CLOCK : Mirage_clock.PCLOCK) (BLOCK : Mirage_block.S) = struc
+            Lwt_result.return ()
+        end >>>= fun () ->
+        let end_bytes = add data_offset end_bytes in
+        let start_bytes = add data_offset offset in
+        let sector_size = of_int t.info.sector_size in
+        let start_sector_offset = rem start_bytes sector_size in
+        let end_sector = div (add end_bytes sector_size) sector_size in
+        let last_sector_offset = rem end_bytes sector_size in
+        (* allocate a buffer for what we need to write, and blit in data and slack
+           at first and last sector. *)
+        let data' =
+          let len =
+            add start_sector_offset
+              (add (of_int (String.length data))
+                 (if last_sector_offset = 0L then
+                    0L
+                  else
+                    sub sector_size last_sector_offset))
+          in
+          Cstruct.create (to_int len)
+        in
+        Cstruct.blit_from_string data 0 data'
+          (to_int start_sector_offset) (String.length data);
+        read_partial_sector t (div start_bytes sector_size) data'
+          ~offset:0L ~length:start_sector_offset >>>= fun () ->
+        let last_sector =
+          Cstruct.sub data' (Cstruct.length data' - t.info.sector_size)
+            t.info.sector_size
+        in
+        read_partial_sector t (pred end_sector) last_sector
+          ~offset:last_sector_offset
+          ~length:(sub sector_size last_sector_offset) >>>= fun () ->
+        (* XXX: this is to work around limitations in some block implementations *)
+        let data' =
+          List.init (Cstruct.length data' / t.info.sector_size)
+            (fun sector ->
+               Cstruct.sub data' (sector * t.info.sector_size) t.info.sector_size)
+        in
+        write t (div start_bytes sector_size) data')
 
-  let allocate _ _ ?last_modified:_ _ =
-    Lwt.return (Error `Append_only)
+  let allocate t key ?last_modified size =
+    Lwt_mutex.with_lock t.write_lock (fun () ->
+        let ( >>>= ) = Lwt_result.bind in
+        let r =
+          let ( let* ) = Result.bind in
+          (* XXX: map `Entry_already_exists to `Append_only ?! *)
+          let* () = is_safe_to_set t key in
+          let hdr = header_of_key ?last_modified key (Optint.Int63.to_int size) in
+          let space_needed = space_needed hdr in
+          let* () = if free t >= space_needed then Ok () else Error `No_space in
+          Ok (hdr, space_needed)
+        in
+        Lwt.return r >>>= fun (hdr, space_needed) ->
+        let open Int64 in
+        let sector_size = of_int t.info.Mirage_block.sector_size in
+        let header_start_bytes =
+          sub t.end_of_archive (of_int (2 * Tar.Header.length))
+        in
+        (* we don't have to zero the block before end_of_archive as it is
+           already zero due to the terminating double zero block (sentinel) *)
+        let to_zero_start_bytes = t.end_of_archive in
+        (* space_needed - 1 block + sentinel = space_needed + 1 block *)
```

Comment:
Should these comments be made clearer? It's an optimization that might make the code less readable - is it worth it?